feat(indexing-language): don't annotate if text is likely binary data

bjorncs opened the pull request:
Skip tokenization if it's a reindexing operation with some heavily throttled logging. If not, fail hard with an exception.

Extended the previous PR with detection of reindexing operations (https://github.com/vespa-engine/vespa/pull/34398). The threshold are the same as before.

FYI @bratseth @hmusum

bjorncs committed:
feat(indexing-language): don't annotate if text is likely binary data

Skip tokenization if it's a reindexing operation with some heavily throttled logging.
If not, fail hard with an exception.

Extended the previous PR with detection of reindexing operations (https://github.com/vespa-engine/vespa/pull/34398).
The threshold are the same as before.

This reverts commit 0ea98903253641351f3044fcc3b3ca04d1490b3f.

bjorncs committed:
The special string is a prefix, not the complete value

bratseth commented:
Great!

The schema/il-script parsers (there are two right now) must also be extended to parse the new tokenization parameters (max-replacement-characters-ratio, and max-replacement-characters). You can find them by searching for "max-token-length".

vekterli reviewed (dismissed):
:+1: with the caveat emptor that the docproc changes should be verified by someone rather more well traveled in those lands than me

arnej27959 commented on docprocs/src/test/java/com/yahoo/docprocs/indexing/DocumentScriptTestCase.java:
```diff
@@ -347,7 +347,7 @@ private static void assertFail(String expectedException, DocumentUpdate update)
 
     private static Document execute(Document document) {
-        return newScript(document.getDataType()).execute(new FieldValuesFactory(), document);
+        return newScript(document.getDataType()).execute(new FieldValuesFactory(), document, isReindexingOperation(input));
```
```suggestion
        return newScript(document.getDataType()).execute(new FieldValuesFactory(), document, false);
```

arnej27959 reviewed (dismissed):
a couple of test methods needs fixing, rest looks good to me

bjorncs committed:
chore: change defaults to conservative values

A ratio of 30% will typically only affect strings that are binary data
only, as a large percentage of binary data will be valid UTF8 codepoints.

bjorncs committed:
chore: improve method name and description

bjorncs committed:
fix: address compilation error in `docprocs` test

bjorncs commented:
I've resolved the requested changes from @arnej27959 / @vekterli and increased the default thresholds to more conservative values. 

@bratseth Imho these settings should rather be global instead of field specific. We can make it easier to override later - if so we should generalize the threshold names (not tie them to replacement characters only). For now the generic config override mechanism should suffice.

bjorncs committed:
feat: temporary backdoor through environment variable to disable binary check

bjorncs committed:
chore: cache value and allow 'false' as value

vekterli reviewed (approved):
:+1:
